### ISSUE TITLE
The shape of embedding issue

### ISSUE THREAD

**Issue by WhiteNan** [MAGICS-LAB/DNABERT_2#121]
Thank you for publishing DNABert2.
I found that the feature lengths obtained by inputting multiple sequences of the same length into DNABert2 are not the same.I would like to know if you have encountered such a problem in downstream tasks, and how should this situation be handled? Is it filled to the same length?
Looking forward to your reply.

**Comment by Zhihan1996** [comment]
Can you share the codes you used to generate the embedding? I guess you did not pad sequences into batches before sending to the model.

**Comment by WhiteNan** [comment]
Thanks for your reply.Here are my codes to generate the embeddings:
labels,sequences = read_tsv("/dataset/"+file_name+".tsv")

device = torch.device('cuda:0' if torch.cuda.is_available() else 'cpu')

all_embeddings = []

tokenizer = AutoTokenizer.from_pretrained("/DNAModel/DNABERT/DNABERT-2-117M",trust_remote_code=True)
model = AutoModel.from_pretrained("/DNAModel/DNABERT/DNABERT-2-117M",trust_remote_code=True).to(device)

with torch.no_grad():
    for seq in sequences:
        inputs = tokenizer(seq, return_tensors='pt')["input_ids"].to(device)
        hidden_states = model(inputs)[0]  # [1, sequence_length, 768]

        trimmed_embeddings = hidden_states[:, 1:-1, :]
        # mean_embeddings = torch.mean(trimmed_embeddings, dim=1)
        # all_embeddings.append(mean_embeddings)
        del hidden_states
        del inputs
        torch.cuda.empty_cache()





------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "MAGICS-LAB/DNABERT_2"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2024年10月28日(星期一) 下午3:52
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [MAGICS-LAB/DNABERT_2] The shape of embedding issue (Issue #121)





 
Can you share the codes you used to generate the embedding? I guess you did not pad sequences into batches before sending to the model.

**Comment by Zhihan1996** [comment]
Sure. In your code, you generate embedding for each sequence separately. So the tokenizer will handle each sequence separately and do not pad them into the same size. If you only wanna the sequence embedding, you can average each output separately along the sequence length dimension to get a fixed-size embedding for each sequence. Otherwise, please refer to `line 143` of `https://github.com/MAGICS-LAB/DNABERT_S/blob/main/evaluate/utils.py` for an example.

**Comment by WhiteNan** [comment]
I don't want the sequence embedding.According to your suggestion, what input format should I use to ensure that the length of the obtained features is the same? Can the obtained feature length be manually set?




------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "MAGICS-LAB/DNABERT_2"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2024年10月28日(星期一) 下午4:03
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [MAGICS-LAB/DNABERT_2] The shape of embedding issue (Issue #121)





 
Sure. In your code, you generate embedding for each sequence separately. So the tokenizer will handle each sequence separately and do not pad them into the same size. If you only wanna the sequence embedding, you can average each output separately along the sequence length dimension to get a fixed-size embedding for each sequence. Otherwise, please refer to line 143 of https://github.com/MAGICS-LAB/DNABERT_S/blob/main/evaluate/utils.py for an example.
